std/machines/hash/poseidon_gl_memory.asm

```diff
@@ -152,16 +152,17 @@ machine PoseidonGLMemory(mem: Memory, split_gl: SplitGL) with
     pol commit output[OUTPUT_SIZE];
 
     // Add round constants
-    let a: expr[STATE_SIZE] = array::zip(state, C, |state, C| state + C);
+    // TODO should these be intermediate?
+    let a = array::zip(state, C, |state, C| state + C);
 
     // Compute S-Boxes (x^7)
-    let x2: expr[STATE_SIZE] = array::map(a, |a| a * a);
-    let x4: expr[STATE_SIZE] = array::map(x2, |x2| x2 * x2);
-    let x6: expr[STATE_SIZE] = array::zip(x4, x2, |x4, x2| x4 * x2);
-    let x7: expr[STATE_SIZE] = array::zip(x6, a, |x6, a| x6 * a);
+    let x2: inter[STATE_SIZE] = array::map(a, |a| a * a);
+    let x4: inter[STATE_SIZE] = array::map(x2, |x2| x2 * x2);
+    let x6: inter[STATE_SIZE] = array::zip(x4, x2, |x4, x2| x4 * x2);
+    let x7: inter[STATE_SIZE] = array::zip(x6, a, |x6, a| x6 * a);
 
     // Apply S-Boxes on the first element and otherwise if it is a full round.
-    let b: expr[STATE_SIZE] = array::new(STATE_SIZE, |i| if i == 0 {
+    let b: expr[] = array::new(STATE_SIZE, |i| if i == 0 {
```

Comment:
is `expr[]` needed here?

Reply:
Yes, because fixed-size arrays is not a type the type checker understands (we would need implicit conversion for this to work nicely). Fixed-size arrays are essentially only useful for column types.